File: my-folder/0213-house-robber-ii/solution.cpp
class Solution {
public:
    unordered_map<int,int> cache;
    int rob(vector<int>& nums) {
        if (nums.size()==0) return 0;
        if (nums.size()==1) return nums[0];
        cache.clear();
        int max1 = dp(nums,0,nums.size()-2);
        cache.clear();
        int max2 = dp(nums,1,nums.size()-1);
        return max(max1,max2);
    }
    int dp(vector<int>& nums,int start,int end)
    {
        if (start>end) return 0;
        if (cache.find(start)!=cache.end()) return cache[start];
        // dont rob the house
        int dont_rob = dp(nums,start+1,end);
        int rob_house = nums[start] + dp(nums,start+2,end);
        cache[start] = max(dont_rob,rob_house);
        return cache[start];

    }
};

File: my-folder/0213-house-robber-ii/solution.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n==1: return nums[0]

        memo = {}
        # solve assuming you rob house 1
        def solve1(i):
            if i>=n-1: return 0
            if i in memo: return memo[i]
            dont_rob = solve1(i+1)
            rob = nums[i] + solve1(i+2)
            memo[i] = max(dont_rob,rob)
            return memo[i]

        memo2 = {}
        def solve2(i):
            if i>=n: return 0
            if i in memo2: return memo2[i]
            dont_rob = solve2(i+1)
            rob = nums[i] + solve2(i+2)
            memo2[i] = max(rob,dont_rob)
            return memo2[i]

        ans = max(solve1(0), solve2(1) )
        return ans


        

